invenio_accounts/models.py

```diff
@@ -463,3 +468,153 @@ def delete_by_user(cls, method, user):
+
+    __tablename__ = "accounts_domain_org"
+
+    id = db.Column(db.Integer(), primary_key=True, autoincrement=True)
+
+    pid = db.Column(db.String(255), unique=True, nullable=True)
+    """Persistent identifier for organisation."""
+
+    name = db.Column(db.String(255), nullable=False)
+    """Name of organisation."""
+
+    json = db.Column(
+        json_field,
+        default=lambda: dict(),
+        nullable=False,
+    )
+    """Store additional metadata about the organisation."""
+
+    parent_id = db.Column(
+        db.Integer(), db.ForeignKey("accounts_domain_org.id"), nullable=True
+    )
+    """Link to parent organisation."""
+
+    parent = db.relationship("DomainOrg", remote_side=[id])
+    """Relationship to parent."""
+
+    domains = db.relationship("Domain", back_populates="org")
+    """Relationship to domains for this organisation."""
+
+    @classmethod
+    def create(cls, pid, name, json=None, parent=None):
+        """Create a domain organisation."""
+        obj = cls(pid=pid, name=name, json=json or {}, parent=parent)
+        db.session.add(obj)
+        return obj
+
+
+class DomainCategory(db.Model):
+    """Model for storing different domain categories."""
+
+    __tablename__ = "accounts_domain_category"
+
+    id = db.Column(db.Integer(), primary_key=True, autoincrement=True)
+
+    label = db.Column(db.String(255))
+
+    @classmethod
+    def create(cls, label):
+        """Create a new domain category."""
+        obj = cls(label=label)
+        db.session.add(obj)
+        return obj
+
+    @classmethod
+    def get(cls, label):
+        """Get a domain category."""
+        return cls.query.filter_by(label=label).one_or_none()
+
+
+class Domain(db.Model, Timestamp):
+    """User domains model."""
+
+    __tablename__ = "accounts_domains"
+
+    id = db.Column(db.Integer(), primary_key=True, autoincrement=True)
+    """Domain ID"""
+
+    _domain = db.Column("domain", db.String(255), unique=True, nullable=False)
+    """Domain name."""
+
+    tld = db.Column(db.String(255), nullable=False)
+    """Top-level domain."""
+
+    status = db.Column(db.Enum(DomainStatus), default=DomainStatus.new, nullable=False)
+    """Status of domain.
+
+    Use to control possibility and capability of users registering with this domain.
+    """
+
+    flagged = db.Column(db.Boolean(), default=False, nullable=False)
+    """Flag domain - used by automatic processes to flag domain."""
+
+    flagged_source = db.Column(db.String(255), default="", nullable=False)
+    """Source of flag."""
+
+    org_id = db.Column(db.Integer(), db.ForeignKey(DomainOrg.id), nullable=True)
+    """Organisation associated with domain."""
+
+    org = db.relationship("DomainOrg", back_populates="domains")
+
+    # spammer, mail-provider, organisation, company
+    category = db.Column(db.Integer(), db.ForeignKey(DomainCategory.id), nullable=True)
+    """Category of domain."""
+
+    num_users = db.Column(db.Integer(), default=0, nullable=False)
```

Comment:
Will be computed on say daily/weekly basis (not part of PR). Needed for easy analysis of domains and detecting domains with high number of blocked users.